ops/main.py

```diff
@@ -11,529 +11,18 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+"""Support legacy ops.main.main() import."""
 
-"""Main entry point to the framework.
-
-Note that this module is callable, and calls the :func:`ops.main.main` function.
-This is so that :code:`import ops` followed by :code:`ops.main(MyCharm)` works
-as expected.
-"""
-
-import logging
-import os
-import shutil
-import subprocess
-import sys
 import warnings
-from pathlib import Path
-from typing import Any, Dict, List, Optional, Tuple, Type, Union, cast
+from typing import Optional, Type
 
 import ops.charm
-import ops.framework
-import ops.model
-import ops.storage
-from ops.charm import CharmMeta
-from ops.jujucontext import _JujuContext
-from ops.log import setup_root_logging
-
-CHARM_STATE_FILE = '.unit-state.db'
```

Comment:
Is this still importable as `from ops.main import CHARM_STATE_FILE`? It doesn't seem like it:

```python
>>> from ops.main import CHARM_STATE_FILE
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'CHARM_STATE_FILE' from 'ops.main' (/home/tameyer/scratch/dima-operator/ops/main.py)
```

Reply:
Good catch, jhack does this:

```
from ops.main import (
15 | CHARM_STATE_FILE,
16 | CharmMeta,
```

And Scenario does this.

Let's agree on Matrix if we re-export some symbols and/or what symbol set to reexport.

Reply:
For what it's worth, I think it's ok to break Scenario (as long as we also fix it before we do an ops release with the breaking change, like we did with `_JujuContext`) but it is a warning that other things might be doing this too.

It seems unlikely that a charm would have any reason to use `CHARM_STATE_FILE` or `setup_root_logging`, but I can definitely see how someone could legitimately be wanting to use `CharmMeta` and they type `ops.main` and get it autocompleted from there.

My preference is to be overly cautious by being maximally backwards compatible (except for silly things like `from ops.main import logging`). If the rest of the team feels like we only need to do that if we can find an example where it would actually break a charm, I can live with that.

Reply:
I'm thinking it's easier to re-export a set of symbols with a deprecation comment and/or proxy function.

Release that, and take our time to update 2 major scenario releases, jhack, etc;

Later on we can clean up those deprecated bits. WDYT?

Reply:
Actually, even that's not entirely enough, consider this jhack code:

```py
import ops.main
--
182 |  
183 | ops.main.setup_root_logging = _patch_logger
```

Reply:
@PietroPasotti a penny for your thoughts!

Reply:
I think we can just import the 2 or 3 symbols from `_main` here, and include a comment about why. Probably not worth trying to formally deprecate or have a deprecation warning.

Reply:
I've decided to re-export the set of symbols that Scenario 6 uses specifically.

Rationale: many charms use Scenario and we shouldn't potentially break a whole bunch of charms if they pin dependencies asymmetrically (compatible ops, exact scenario); and to decouple Scenario 6 update from this PR in calendar time.